proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+Options 2 and 3 have similar characteristics: They both express that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. Option 2 is likely cheaper, as the argument with `ref/out SwiftError<T>` type can easily be associated with the error register. With Option 2 though, we'd have to determine the behavior if multiple parameters used the `SwiftError` type (likely throw an exception). Option 3 provides a more functional-programming-style API that is more intuitive, but likely more expensive to implement.
+
+Option 4 provides an alternative approach. As the "Error Register" is always register-sized, we can use runtime-supported helper functions to stash away and retrieve the error register. Unlike options 2 or 3, we don't need to do any signature manipulation as the concept of "this Swift call can return an error" is not represented in the signature. Responsibility to convert the returned error value into a .NET type, such as an exception, would fall to the projection tooling. Since option 4 does not express whether or not the target function throws at the signature level, the JIT/AOT compilers would always need to emit the calls to the helpers when compiling calls to and from Swift code. If the projection of Swift types into .NET will always use exception and not pass Swift errors as the error types directly, then Option 4 reduces the design burden on the runtime teams by removing the type that would only be used at the lowest level of the Swift/.NET interop.
+
+##### Tuples
+
+In the Swift language, tuples are "exploded" out in the calling convention. Each element of a tuple is passed as a separate parameter. C# has its own language-level tuple feature in the ValueTuple family of types. The Swift/.NET interop story could choose to automatically handle tuple types in a similar way for a `CallConvSwift` function call. However, processing the tuple types in the JIT/AOT compilers is complicated and expensive. It would be much cheaper to defer this work upstack to the language projection, which can split out the tuple into individual parameters to pass to the underlying function. The runtime and libraries teams could add analyzers to detect value-tuples being passed to `CallConvSwift` functions at compile time to help avoid pits of failure.
+
+##### Automatic Reference Counting and Lifetime Management
+
+Swift has a very strongly-defined lifetime and ownership model. This model is specified in the Swift ABI and is similar to Objective-C's ARC (Automatic Reference Counting) system. The Binding Tools for Swift tooling handles these explicit lifetime semantics with some generated Swift code. In the new Swift/.NET interop, management of these lifetime semantics will be done by the Swift projection and not by the raw calling-convention support.
```

Comment:
How is this implemented in Objective-C interop? Is it the responsibility of projection tools and not the runtime? If we consider future extensibility, it may be beneficial to handle it at the runtime level.

Reply:
> How is this implemented in Objective-C interop? 

There is a complex dance between the runtime/GC and the code shipped out of macios. See https://github.com/dotnet/designs/blob/main/accepted/2021/objectivec-interop.md and https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.objectivec.